rust/template/differential_datalog/src/program/worker.rs

```diff
@@ -428,10 +442,36 @@ impl<'a> DDlogWorker<'a> {
                     HashMap::with_capacity_and_hasher(program.nodes.len(), FnvBuildHasher::default());
             let mut arrangements = FnvHashMap::default();
 
+            // Create an `Enabled` relation used to enforce the dataflow termination in the
+            // presence of delayed relations.  A delayed relation can potentially generate an
+            // infinite sequence of outputs for all future timestamps, preventing the
+            // differential dataflow from terminating.  DD can only terminate cleanly when there
+            // is no data in flight, and will keep spinning forever if new records keep getting
+            // injected in the dataflow.  To prevent this scenario, we join all delayed relations
+            // with `Enabled`.  `Enabled` contains a single record (an empty tuple) as long as
+            // the program is running.  We retract this record on shutdown, hopefully enforcing
+            // quiescing the dataflow.
+            let (enabled_session, enabled_collection) = outer.new_collection::<(),Weight>();
```

Comment:
Is this overhead?

Reply:
yep, but a small one, as `Enabled` contains exactly one record, and we only pay it if there are delayed relations in the program.  For the one benchmark we have for this now, it does not seem to make any difference.